core/chain_indexer.go

```diff
@@ -0,0 +1,296 @@
+// Further child ChainIndexers can be added which use the output of the parent section
+// indexer. These child indexers receive new head notifications only after an entire section
+// has been finished or in case of rollbacks that might affect already finished sections.
+type ChainIndexer struct {
+	db                                          ethdb.Database
+	validSectionsKey, sectionHeadKeyPrefix      []byte
+	backend                                     ChainIndex
+	sectionSize, confirmReq                     uint64
+	stop                                        chan struct{}
+	lock                                        sync.Mutex
+	procWait                                    time.Duration
+	tryUpdate                                   chan struct{}
+	stored, targetCount, calcIdx, lastForwarded uint64
+	updating                                    bool
+	children                                    []*ChainIndexer
+}
+
+// ChainIndex interface is a backend for the indexer doing the actual post-processing job
+type ChainIndex interface {
+	Reset(section uint64)           // start processing a new section
+	Process(header *types.Header)   // process a single block (called for each block in the section)
+	Commit(db ethdb.Database) error // do some more processing if necessary and store the results in the database
+	UpdateMsg(done, all uint64)     // print a progress update message if necessary (only called when multiple sections need to be processed)
+}
+
+// NewChainIndexer creates a new  ChainIndexer
+//  db:				database where the index of available processed sections is stored (the index is stored by the
+//                  indexer, the actual processed chain data is stored by the backend)
+//  dbKey:			key prefix where the index is stored
+//  backend:		an implementation of ChainIndex
+//  sectionSize:	the size of processable sections
+//  confirmReq:		required number of confirmation blocks before a new section is being processed
+//  procWait:		waiting time between processing sections (simple way of limiting the resource usage of a db upgrade)
+//  stop:		    quit channel
+func NewChainIndexer(db ethdb.Database, dbKey []byte, backend ChainIndex, sectionSize, confirmReq uint64, procWait time.Duration, stop chan struct{}) *ChainIndexer {
+	c := &ChainIndexer{
+		db:                   db,
+		validSectionsKey:     append(dbKey, []byte("-count")...),
+		sectionHeadKeyPrefix: append(dbKey, []byte("-shead")...),
+		backend:              backend,
+		sectionSize:          sectionSize,
+		confirmReq:           confirmReq,
+		tryUpdate:            make(chan struct{}, 1),
+		stop:                 stop,
+		procWait:             procWait,
+	}
+	c.stored = c.getValidSections()
+	go c.updateLoop()
+	return c
+}
+
+// updateLoop is the main event loop of the indexer
+func (c *ChainIndexer) updateLoop() {
+	updateMsg := false
+
+	for {
+		select {
+		case <-c.stop:
+			return
+		case <-c.tryUpdate:
+			c.lock.Lock()
```

Comment:
I've given it some thought but I think it's better to keep it like this. The update loop is a slow loop that blocks while processing sections so I don't want to pass all new head events, only send a "wake up" signal to tryUpdate channel if necessary (when updating flag is false and there is something new to update). Also, it's not ideal to process all new head events sequentially because some new sections might have already been rolled back by the time the update loop gets there.